# Python Crash Course
## Compbio Camp - February 2019
### Author: Chris Porras


# <font color=blue\>*YOUR NAME HERE*</font>

**Welcome to Python!** <br>
This will be a brief crash course of the basic computational techniques that we will be using throughout this camp. 

### Some quick vocabulary:

> * __Jupyter Notebook__: a computational environment that allows us to neatly format and keep track of our Python code in <font color=blue>"notebooks"</font>. <br>
<br>
* __Cell__: the darker grey space below denoted by <font color=blue>"  In [ &ensp; ]  :  "</font>. Each cell is a convenient container for us to write a segment of Python code in. We can run the code inside of a cell using the `ctrl + enter` or `shift + enter` keyboard shortcuts. Cells are used to separate parts of our code that we want to run separately from one another.

## Let's jump into it!

### 1. Variable assignments and arithmetic

> Variables hold information in the form of integers, decimals (called ___"floats"___), words (called ___"strings"___ ), and others. We can assign values to variables with the " __=__ " assignment operator. Values assigned to variables are called ___"parameters"___ or ___"constants"___ and are stored in memory to be "called" by their variable names in other operations. 

In [44]:
## P.S. We can write "comments" in cells using the '#'. 
# Writing comments is useful for annotating our code as we write it. 
# Comments are ignored when the code is run.

####### Example #######
a = 5
b = 0.5
c = "dogs" # strings can be written with either double "" or single '' quotes.

> We can do standard arithmetic with variables and the __(+  ,  -  ,  *  ,  /  )__ operators. 

In [24]:
a + b

5.5

In [25]:
a - b

4.5

In [45]:
a * b

2.5

In [27]:
a / b 

10.0

> We can write exponents with the " __**__ " operator.

In [43]:
####### Example #######
2 ** 3

8

> __Question__: What do you think will happen when we perform the following operation? <br>
`a + c`

In [46]:
####### YOUR CODE HERE #######


> __Answer__: we get a `TypeError` when we try to perform arithmetic with an integer and a string. This is how Python lets us know that we're asking it to do something that doesn't make sense. <br>
Was your prediction correct?

> When performing assignments, the variable names don't really matter as long as you are consistent with them. It's good coding practice to use descriptive but brief names.

In [47]:
####### Example #######
hrs_in_day = 24
num_days = 7

hrs_in_week = hrs_in_day * num_days

hrs_in_week

168

#### Practice:

> Write code using variable assignments to show the results of the following: <br>
1. $\frac{3}{2}\times 5$
2. $10\times 0.3$
3. $4^2 - 3$

In [43]:
####### YOUR CODE HERE #######
# 1.


In [44]:
####### YOUR CODE HERE #######
# 2.


In [45]:
####### YOUR CODE HERE #######
# 3.


> __Answers__:
1. 7.5
2. 3.0
3. 13

### 2. Lists

> * __Lists:__ are a useful data type that allows us to combine values into a single form. We can make lists with the " [&emsp;] " operator surrounding a sequence of values to combine. <br><br>
* __Function(&ensp;)__: the (&ensp;) denote functions. Functions can be fed inputs, or ___"arguments"___, and will perform operations and provide an output when run. 

In [21]:
####### Example #######
# We can assign lists to specific names to access the information sequence all at once
odd_nums = [1,3,5,7]

even_nums = [2,4,6,8]

# The print() function returns the values of its inputs. 

# Where before we needed to have separate cells for each single output,
# we can now use print() to show the outputs we want in the same cell. 

print(odd_nums)
print(even_nums)

[1, 3, 5, 7]
[2, 4, 6, 8]
[1, 3, 5, 7, 2, 4, 6, 8]


> __Question__: What do you think will happen when you perform the following operation? <br>
`print(odd_nums + even_nums)`

In [ ]:
####### YOUR CODE HERE #######


> __Answer__: adding lists _doesn't_ work the same way that adding variables does. Instead we combine the lists into larger ones. To perform typical arithmetic with lists we will need to convert them into a new data type that plays nicely with arithmetic operations. <br>
How did this compare to your prediction?

### 3. Building arrays and vector arithmetic in `NumPy`

> * __Array__: A structure that stores a greater amount of information in an organized way (often arranged in rows and columns). The data stored in an array must be of the same type (integers, floats, etc.). Arrays with only one row or one column of data are called ___"1-dimensional"___. <br> <br>

> __NOTE__ : The following cell imports `NumPy`, a useful Python package for manipulating arrays. Please make sure to run that cell before continuing. 

In [2]:
import numpy as np

> Arrays can be thought of as a list of values, or a collection of variables. In the following examples, I will show how we can create arrays using `NumPy` and perform the same arithmetic operations as before.

In [3]:
####### Example #######
my_arr = np.array([2,3,9,1]) # Give the np.array() function a list to create an array from 

arr2 = np.array([2,10,7,22])

print(my_arr)
print(arr2)
print(my_arr + arr2) # Now we can do typical arithmetic

[2 3 9 1]
[ 2 10  7 22]
[ 4 13 16 23]


**Practice:**

> Create the following arrays: <br>
       &#9656;  $\textbf{x}\ = (1,3,5,7,9)$ <br>
       &#9656;  $\textbf{y}\ = (2,4,6,8,10)$<br><br>
> Perform the following operations:
1. $\textbf{x}+\textbf{y}$
2. $\textbf{x}\times\textbf{y}$
3. $2\textbf{x}-\frac{\textbf{y}}{2}$ <br>

Please use `print()` to output your answers in the same cell.

In [48]:
####### YOUR CODE HERE #######


> __Answers__:<br>
1. array(  [ 3,  7, 11, 15, 19]  )
2. array(  [ 2, 12, 30, 56, 90]  )
3. array(  [ 1.,  4.,  7., 10., 13.]  )

### 4. Indexing

> Let's say that instead of performing operations with the entire array, we only want to work with a specific subset of values in the array. *How can we do this?* <br><br>
Python arrays have an underlying identification system for each location in an array. Each location is given an "___index___" which counts from __0__ to the end of the array. We use "__[ 0 ]__" to get the first value in the array.<br><br>
Take a look at the following example:

In [12]:
####### Example #######
price = np.array([2.5,2.68,2.72,2.13,2.14]) 

print(price[0],"is the first value in price")
print(price[1:3],
      "are the second through the the fourth values in price (not including the fourth)")
print(price[-1],"gives the last value in price")

2.5 is the first value in price
[2.68 2.72] are the second through the the fourth values in price (not including the fourth)
2.14 gives the last value in price


> __Question__: How would you use indexing to change the __second__ element of the following array to a __7__?

In [16]:
####### YOUR CODE HERE #######
time = np.array([6.5,3,7.2,8,7.4])


> __Answer__:

In [15]:
time[1] = 7
print(time)

[6.5 7.  7.2 8.  7.4]


> Replacing specified values of an array with other values is hugely useful in scientific programming. Let's work together through the following exercise: 

#### Practice:
1. A population of yeast cells undergoes mitosis every hour, __doubling__ in population size. If there are __10 cells initially__, how many cells will there be after __3 hours__?

In [22]:
# We first create an array of zeros to store our information 
cells = np.array([0,0,0,0]) 

# Next we assign the initial value to the first index of our array
cells[0] = 10

# Finally we double the population size for each step, 
# and using the index of the previous value,
# assign the result to the next index.
cells[1] = cells[0]*2
cells[2] = cells[1]*2
cells[3] = cells[2]*2

# We print our results:
print(cells)
print("After 3 hours, there will be ", cells[3], " cells.")

[10 20 40 80]
After 3 hours, there will be  80  cells.


#### Practice:
1. Chicago is experiencing a flu epidemic. There are now __10,000__ people with the flu, and an __additional 2000__ will become infected each day. After how many days will the population of infected individuals have __exceeded 15,000__?

In [ ]:
####### YOUR CODE HERE #######
# HINT: Use the same method as before, but with the operations described above^^^


> __Answer__: <br>
After 4 days. 

### 5. Loops

> Looking back at the calculations we just made, it sure looks like we need an easier way to perform repetitive tasks. Lucky for us, we can write " __for loops__ " to repeat a section of code however many times we want. 

In [43]:
####### Example #######
# Using for loops with the yeast cells problem looks like this:

# We still want to create the same blank array first
num_hours = 3 # from the yeast problem
cells = np.zeros(num_hours+1) # np.zeros() makes an array of zeros with a length specified by the input.
# the +1 adds a zero for the initial value

# Next, we still assign the initial value to our array
cells[0] = 10

# BUT now we write a for loop to perform the calculation we want:
for i in range(num_hours): 
    #take the current value of cells[i], multiply by 2, and assign that to the next value
    cells[i+1] = cells[i]*2 
    
# We print our results:
print(cells)
print("After 3 hours, there will be ", cells[3], " cells.")

[10. 20. 40. 80.]
After 3 hours, there will be  80.0  cells.


> Let's look more closely at the skeleton of a for loop:

In [44]:
num_reps = 5 # variable that specifies the number of times to repeat the loop

# i is a counting variable that increases by one and goes up to num_reps
# range() tells the loop to start at i = 0 and go to i = (num_reps-1)
for i in range(num_reps): 
    print(i)

## We use this counting " i " to index arrays and perform repetitive operations.  

0
1
2
3
4


#### Practice:
Use for loops and indexing to answer:
1. Every year, the population of mice in a certain environment doubles. Unfortunately, 10 mice will be eaten by snakes each year. There are currently 15 mice. What will happen to the mice population after 5 years?

In [54]:
num_years = 5
mice = np.zeros(num_years+1)
mice[0] = 15

for i in range(num_years):
    mice[i+1] = mice[i]*2 -10
    
print(mice)

[ 15.  20.  30.  50.  90. 170.]


> __Answer__: <br>
There will be 170 mice. 

> __Question__: What happens if you change the initial mouse population to 10? to 5? <br>
Does this make sense?

### 6. Conditional Statements


> IF & ELSE FOR SIMPLE MODEL

### 6. Simple modeling and plotting data

> PREDATION MODEL